File: ios/Packages/PicPacKit/Sources/PicPacAI/ExpectiminimaxAgent.swift
import PicPacCore

public struct SearchAnalysis: Equatable, Sendable {
    public let bestCell: Cell
    public let value: Double
    public let actionValues: [Cell: Double]
    public let nodes: Int
    public let cacheHits: Int
    public let reachedDepth: Int

    public init(
        bestCell: Cell,
        value: Double,
        actionValues: [Cell: Double],
        nodes: Int,
        cacheHits: Int,
        reachedDepth: Int
    ) {
        self.bestCell = bestCell
        self.value = value
        self.actionValues = actionValues
        self.nodes = nodes
        self.cacheHits = cacheHits
        self.reachedDepth = reachedDepth
    }
}

public actor ExpectiminimaxSolver {
    private var cache: [UInt64: Double] = [:]
    private var nodes = 0
    private var cacheHits = 0
    private var deepest = 0
    private var limits = SearchLimits()

    public init() {
        cache.reserveCapacity(40_000)
    }

    public var cachedStateCount: Int { cache.count }

    public func clearCache() {
        cache.removeAll(keepingCapacity: true)
    }

    public func analyze(
        _ state: PicPacDecisionState,
        maxDepth: Int = .max,
        limits: SearchLimits = SearchLimits()
    ) throws -> SearchAnalysis {
        precondition(maxDepth > 0)
        try Task.checkCancellation()
        nodes = 0
        cacheHits = 0
        deepest = 0
        self.limits = limits

        let depth = min(maxDepth, state.legalCells.count)
        var values: [Cell: Double] = [:]
        values.reserveCapacity(state.legalCells.count)
        for cell in PositionEvaluator.ordered(state.legalCells) {
            try Task.checkCancellation()
            values[cell] = try actionValue(state, cell: cell, depth: depth, ply: 0)
        }
        guard let bestValue = values.values.max(),
              let bestCell = PositionEvaluator.ordered(state.legalCells).first(where: {
                  guard let value = values[$0] else { return false }
                  return abs(value - bestValue) < 1e-12
              }) else {
            throw SearchModelError.invalidState
        }
        return SearchAnalysis(
            bestCell: bestCell,
            value: bestValue,
            actionValues: values,
            nodes: nodes,
            cacheHits: cacheHits,
            reachedDepth: deepest
        )
    }

    private func value(_ state: PicPacDecisionState, depth: Int, ply: Int) throws -> Double {
        try checkpoint(ply: ply)
        if depth == 0 { return try PositionEvaluator.boundedValue(state) }

        let cacheKey = key(state, depth: depth)
        if let cached = cache[cacheKey] {
            cacheHits += 1
            return cached
        }

        var best = -Double.infinity
        for cell in PositionEvaluator.ordered(state.legalCells) {
            best = max(best, try actionValue(state, cell: cell, depth: depth, ply: ply))
        }
        cache[cacheKey] = best
        return best
    }

    private func actionValue(
        _ state: PicPacDecisionState,
        cell: Cell,
        depth: Int,
        ply: Int
    ) throws -> Double {
        switch try PublicPicPacSearchModel.place(state, at: cell) {
        case .win:
            return 1
        case .draw:
            return 0
        case let .chance(chance):
            return try -expectedOpponentValue(chance, depth: depth - 1, ply: ply + 1)
        }
    }

    private func expectedOpponentValue(
        _ chance: PicPacChanceState,
        depth: Int,
        ply: Int
    ) throws -> Double {
        let total = Double(chance.total)
        var expected = 0.0
        if chance.remainingX > 0 {
            let next = try PublicPicPacSearchModel.draw(chance, symbol: .x)
            expected += Double(chance.remainingX) / total * (try value(next, depth: depth, ply: ply))
        }
        if chance.remainingO > 0 {
            let next = try PublicPicPacSearchModel.draw(chance, symbol: .o)
            expected += Double(chance.remainingO) / total * (try value(next, depth: depth, ply: ply))
        }
        return expected
    }

    private func checkpoint(ply: Int) throws {
        nodes += 1
        deepest = max(deepest, ply)
        if nodes & 255 == 0 { try Task.checkCancellation() }
        if let nodeBudget = limits.nodeBudget, nodes > nodeBudget {
            throw SearchLimitReached()
        }
        if limits.isPastDeadline {
            throw SearchLimitReached()
        }
    }

    private func key(_ state: PicPacDecisionState, depth: Int) -> UInt64 {
        let held: UInt64 = state.heldSymbol == .x ? 0 : 1
        return UInt64(state.board.code)
            | (held << 15)
            | (UInt64(state.remainingX) << 16)
            | (UInt64(state.remainingO) << 19)
            | (UInt64(depth) << 22)
    }
}

public actor ExpectiminimaxAgent: AiAgent {
    private let configuredDepth: Int
    private let solver: ExpectiminimaxSolver
    private let fallback: any AiAgent

    public init(
        configuredDepth: Int = .max,
        solver: ExpectiminimaxSolver = ExpectiminimaxSolver(),
        fallback: any AiAgent = HeuristicAgent()
    ) {
        precondition(configuredDepth > 0)
        self.configuredDepth = configuredDepth
        self.solver = solver
        self.fallback = fallback
    }

    public func chooseMove(
        for observation: AiObservation,
        limits: SearchLimits = SearchLimits()
    ) async throws -> AiDecision {
        let state = PublicPicPacSearchModel.from(observation)
        let clock = ContinuousClock()
        let start = clock.now
        let analysis: SearchAnalysis
        do {
            analysis = try await solver.analyze(
                state,
                maxDepth: limits.maxDepth ?? configuredDepth,
                limits: limits
            )
        } catch is SearchLimitReached {
            return try await fallback.chooseMove(
                for: observation,
                limits: limits.withoutResourceBounds
            )
        }
        return AiDecision(
            cell: analysis.bestCell,
            diagnostics: .search(
                SearchDiagnostics(
                    nodes: analysis.nodes,
                    cacheHits: analysis.cacheHits,
                    maxDepth: analysis.reachedDepth,
                    elapsedNanoseconds: elapsedNanoseconds(start.duration(to: clock.now)),
                    value: analysis.value
                )
            )
        )
    }
}

private struct SearchLimitReached: Error {}

File: game-ai/src/main/kotlin/com/thevaguebox/picpac/ai/ExpectiminimaxAgent.kt
package com.thevaguebox.picpac.ai

import com.thevaguebox.picpac.core.Cell
import com.thevaguebox.picpac.core.Symbol
import com.thevaguebox.picpac.core.ai.AiAgent
import com.thevaguebox.picpac.core.ai.AiDecision
import com.thevaguebox.picpac.core.ai.AiDiagnostics
import com.thevaguebox.picpac.core.ai.AiObservation
import com.thevaguebox.picpac.core.ai.PicPacChanceState
import com.thevaguebox.picpac.core.ai.PicPacDecisionState
import com.thevaguebox.picpac.core.ai.PublicPicPacSearchModel
import com.thevaguebox.picpac.core.ai.SearchLimits
import com.thevaguebox.picpac.core.ai.SearchTransition
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlin.math.min
import kotlin.system.measureNanoTime

data class SearchAnalysis(
    val bestCell: Cell,
    val value: Double,
    val actionValues: Map<Cell, Double>,
    val nodes: Long,
    val cacheHits: Long,
    val reachedDepth: Int,
)

class ExpectiminimaxSolver {
    private val cache = HashMap<Long, Double>(40_000)
    private var nodes = 0L
    private var cacheHits = 0L
    private var deepest = 0
    private var deadlineNanos: Long? = null
    private var nodeBudget: Int? = null

    suspend fun analyze(
        state: PicPacDecisionState,
        maxDepth: Int = Int.MAX_VALUE,
        limits: SearchLimits = SearchLimits(),
    ): SearchAnalysis {
        nodes = 0
        cacheHits = 0
        deepest = 0
        deadlineNanos = limits.deadlineNanos
        nodeBudget = limits.nodeBudget
        val depth = min(maxDepth, state.legalCells.size)
        val values = linkedMapOf<Cell, Double>()
        for (cell in PositionEvaluator.ordered(state.legalCells)) {
            values[cell] = actionValue(state, cell, depth, 0)
        }
        val bestValue = values.maxOf { it.value }
        val bestCell = values.entries.first { nearlyEqual(it.value, bestValue) }.key
        return SearchAnalysis(bestCell, bestValue, values, nodes, cacheHits, deepest)
    }

    fun clearCache() = cache.clear()
    val cachedStateCount: Int get() = cache.size

    private suspend fun value(state: PicPacDecisionState, depth: Int, ply: Int): Double {
        checkpoint(ply)
        if (depth == 0) return PositionEvaluator.boundedValue(state)
        val key = key(state, depth)
        cache[key]?.let {
            cacheHits++
            return it
        }
        var best = Double.NEGATIVE_INFINITY
        for (cell in PositionEvaluator.ordered(state.legalCells)) {
            val candidate = actionValue(state, cell, depth, ply)
            if (candidate > best) best = candidate
        }
        cache[key] = best
        return best
    }

    private suspend fun actionValue(
        state: PicPacDecisionState,
        cell: Cell,
        depth: Int,
        ply: Int,
    ): Double = when (val transition = PublicPicPacSearchModel.place(state, cell)) {
        SearchTransition.Win -> 1.0
        SearchTransition.Draw -> 0.0
        is SearchTransition.Chance -> -expectedOpponentValue(transition.state, depth - 1, ply + 1)
    }

    private suspend fun expectedOpponentValue(
        chance: PicPacChanceState,
        depth: Int,
        ply: Int,
    ): Double {
        val total = chance.total.toDouble()
        var expected = 0.0
        if (chance.remainingX > 0) {
            expected += chance.remainingX / total * value(
                PublicPicPacSearchModel.draw(chance, Symbol.X), depth, ply,
            )
        }
        if (chance.remainingO > 0) {
            expected += chance.remainingO / total * value(
                PublicPicPacSearchModel.draw(chance, Symbol.O), depth, ply,
            )
        }
        return expected
    }

    private suspend fun checkpoint(ply: Int) {
        nodes++
        if (ply > deepest) deepest = ply
        if ((nodes and 255L) == 0L) currentCoroutineContext().ensureActive()
        if (nodeBudget?.let { nodes > it } == true) throw SearchLimitReached()
        if (deadlineNanos?.let { System.nanoTime() >= it } == true) throw SearchLimitReached()
    }

    private fun key(state: PicPacDecisionState, depth: Int): Long {
        val held = if (state.heldSymbol == Symbol.X) 0L else 1L
        return state.board.code.toLong() or
            (held shl 15) or
            (state.remainingX.toLong() shl 16) or
            (state.remainingO.toLong() shl 19) or
            (depth.toLong() shl 22)
    }

    private fun nearlyEqual(a: Double, b: Double) = kotlin.math.abs(a - b) < 1e-12
}

class ExpectiminimaxAgent(
    private val configuredDepth: Int = Int.MAX_VALUE,
    private val solver: ExpectiminimaxSolver = ExpectiminimaxSolver(),
    private val fallback: AiAgent = HeuristicAgent(),
) : AiAgent {
    override suspend fun chooseMove(observation: AiObservation, limits: SearchLimits): AiDecision {
        val state = PublicPicPacSearchModel.from(observation)
        var analysis: SearchAnalysis? = null
        val elapsed = measureNanoTime {
            analysis = try {
                solver.analyze(state, limits.maxDepth ?: configuredDepth, limits)
            } catch (_: SearchLimitReached) {
                null
            }
        }
        val result = analysis ?: return fallback.chooseMove(observation, limits.copy(nodeBudget = null, deadlineNanos = null))
        return AiDecision(
            cell = result.bestCell,
            diagnostics = AiDiagnostics.Search(
                nodes = result.nodes,
                cacheHits = result.cacheHits,
                maxDepth = result.reachedDepth,
                elapsedNanos = elapsed,
                value = result.value,
            ),
        )
    }
}

private class SearchLimitReached : RuntimeException(null, null, false, false)
